ecc/bls12-377/fr/gkr/registry.go

```diff
@@ -0,0 +1,301 @@
+	return func(settings *registerGateSettings) {
+		settings.noDegreeVerification = true
+		settings.degree = degree
+	}
+}
+
+// WithDegree sets the degree of the gate. RegisterGate will return an error if the degree is not correct.
+func WithDegree(degree int) RegisterGateOption {
+	return func(settings *registerGateSettings) {
+		settings.degree = degree
+	}
+}
+
+// isAdditive returns whether x_i occurs only in a monomial of total degree 1 in f
+func (f GateFunction) isAdditive(i, nbIn int) bool {
+	// fix all variables except the i-th one at random points
+	// pick random value x1 for the i-th variable
+	// check if f(-, 0, -) + f(-, 2*x1, -) = 2*f(-, x1, -)
+	x := make(fr.Vector, nbIn)
+	x.MustSetRandom()
+	x0 := x[i]
+	x[i].SetZero()
+	in := slices.Clone(x)
+	y0 := f(in...)
+
+	x[i] = x0
+	copy(in, x)
+	y1 := f(in...)
+
+	x[i].Double(&x[i])
+	copy(in, x)
+	y2 := f(in...)
+
+	y2.Sub(&y2, &y1)
+	y1.Sub(&y1, &y0)
+
+	if !y2.Equal(&y1) {
+		return false // not linear
+	}
+
+	// check if the coefficient of x_i is nonzero and independent of the other variables (so that we know it is ALWAYS nonzero)
+	if y1.IsZero() { // f(-, x1, -) = f(-, 0, -), so the coefficient of x_i is 0
+		return false
+	}
+
+	// compute the slope with another assignment for the other variables
+	x.MustSetRandom()
+	x[i].SetZero()
+	copy(in, x)
+	y0 = f(in...)
+
+	x[i] = x0
+	copy(in, x)
+	y1 = f(in...)
+
+	y1.Sub(&y1, &y0)
+
+	return y1.Equal(&y2)
+}
+
+// fitPoly tries to fit a polynomial of degree less than degreeBound to f.
+// degreeBound must be a power of 2.
+// It returns the polynomial if successful, nil otherwise
+func (f GateFunction) fitPoly(nbIn int, degreeBound uint64) polynomial.Polynomial {
+	// turn f univariate by defining p(x) as f(x, x, ..., x)
+	fIn := make([]fr.Element, nbIn)
+	p := make(polynomial.Polynomial, degreeBound)
+	domain := fft.NewDomain(degreeBound)
+	// evaluate p on the unit circle (first filling p with evaluations rather than coefficients)
+	x := fr.One()
+	for i := range p {
+		for j := range fIn {
+			fIn[j] = x
+		}
+		p[i] = f(fIn...)
+
+		x.Mul(&x, &domain.Generator)
+	}
+
+	// obtain p's coefficients
+	domain.FFTInverse(p, fft.DIF)
+	fft.BitReverse(p)
+
+	// check if p is equal to f. This not being the case means that f is of a degree higher than degreeBound
+	fIn[0].MustSetRandom()
+	for i := range fIn {
+		fIn[i] = fIn[0]
+	}
+	pAt := p.Eval(&fIn[0])
+	fAt := f(fIn...)
+	if !pAt.Equal(&fAt) {
+		return nil
+	}
+
+	// trim p
+	lastNonZero := len(p) - 1
+	for lastNonZero >= 0 && p[lastNonZero].IsZero() {
+		lastNonZero--
+	}
+	return p[:lastNonZero+1]
+}
+
+// FindDegree returns the degree of the gate function, or -1 if it fails.
+// Failure could be due to the degree being higher than max or the function not being a polynomial at all.
+func (f GateFunction) FindDegree(max, nbIn int) int {
+	bound := uint64(max) + 1
+	for degreeBound := uint64(4); degreeBound <= bound; degreeBound *= 2 {
```

Comment:
Maybe bigger increase factor? With 2 we most most definitely run the loop mulltiple times. But for any practical gates if we would have 8-16 etc then would have a single loop.

Reply:
Makes sense. What about the initial value?

Reply:
Yup, makes sense to me.